subsys/usb/device/class/bt_h4.c

```diff
@@ -93,25 +97,92 @@ static struct usb_ep_cfg_data bt_h4_ep_data[] = {
 	},
 };
 
+static uint16_t hci_pkt_get_len(struct net_buf *buf,
+				const uint8_t *data, size_t size)
+{
+	uint16_t len = 0;
+	size_t hdr_len = 0;
+
+	switch (bt_buf_get_type(buf)) {
+	case BT_BUF_CMD: {
+		struct bt_hci_cmd_hdr *cmd_hdr;
+
+		hdr_len = sizeof(*cmd_hdr);
+		cmd_hdr = (struct bt_hci_cmd_hdr *)data;
+		len = cmd_hdr->param_len + hdr_len;
+		break;
+	}
+	case BT_BUF_ACL_OUT: {
+		struct bt_hci_acl_hdr *acl_hdr;
+
+		hdr_len = sizeof(*acl_hdr);
+		acl_hdr = (struct bt_hci_acl_hdr *)data;
+		len = sys_le16_to_cpu(acl_hdr->len) + hdr_len;
+		break;
+	}
+	case BT_BUF_ISO_OUT: {
+		struct bt_hci_iso_hdr *iso_hdr;
+
+		hdr_len = sizeof(*iso_hdr);
+		iso_hdr = (struct bt_hci_iso_hdr *)data;
+		len = bt_iso_hdr_len(sys_le16_to_cpu(iso_hdr->len)) + hdr_len;
+		break;
+	}
+	default:
+		LOG_ERR("Unknown bt buffer type");
+		return 0;
+	}
+
+	return (size < hdr_len) ? 0 : len;
+}
+
 static void bt_h4_read(uint8_t ep, int size, void *priv)
 {
+	static struct net_buf *buf;
 	static uint8_t data[USB_MAX_FS_BULK_MPS];
+	static uint16_t pkt_len;
 
-	if (size > 0) {
-		struct net_buf *buf;
+	if (size == 0) {
+		goto restart_out_transfer;
+	}
 
+	if (buf == NULL) {
 		buf = bt_buf_get_tx(BT_BUF_H4, K_FOREVER, data, size);
 		if (!buf) {
-			LOG_ERR("Cannot get free TX buffer\n");
-			return;
+			LOG_ERR("Failed to allocate buffer");
+			goto restart_out_transfer;
+		}
+
+		pkt_len = hci_pkt_get_len(buf, &data[1], size - 1);
+		LOG_DBG("pkt_len %u, chunk %u", pkt_len, size);
+	} else {
+		if (net_buf_tailroom(buf) < size) {
+			LOG_ERR("Buffer tailroom too small");
+			net_buf_unref(buf);
+			buf = NULL;
+			goto restart_out_transfer;
 		}
 
+		/*
+		 * Take over the next chunk if HCI packet is
+		 * larger than USB_MAX_FS_BULK_MPS.
+		 */
+		net_buf_add_mem(buf, data, size);
+		LOG_DBG("len %u, chunk %u", buf->len, size);
+	}
+
+	if (buf != NULL && pkt_len == buf->len) {
 		net_buf_put(&rx_queue, buf);
+		LOG_DBG("put");
+		buf = NULL;
+		pkt_len = 0;
 	}
 
+restart_out_transfer:
 	/* Start a new read transfer */
 	usb_transfer(bt_h4_ep_data[BT_H4_OUT_EP_IDX].ep_addr, data,
-		     USB_MAX_FS_BULK_MPS, USB_TRANS_READ, bt_h4_read, NULL);
+		     USB_MAX_FS_BULK_MPS, USB_TRANS_READ | USB_TRANS_NO_ZLP,
```

Comment:
Why? `USB_TRANS_NO_ZLP` has no real meaning from device perspective for OUT transfers. Or do you expect this to do something? If host writes ZLP, then it'll still be received by `bt_h4_read`. Since the buffer length is Max Packet Size, the `bt_h4_read` will be called on every received OUT DATA packet.

Reply:
It might be a good comment, does the same apply for: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/usb/device/class/bluetooth.c#L300?

Reply:
This is copy&paste from bluetooth.c and is there for reasons.

Reply:
`acl_read_cb` is also OUT transfer, just like `bt_h4_read`. @jfischer-no what reasons?

Reply:
> `acl_read_cb` is also OUT transfer, just like `bt_h4_read`. @jfischer-no what reasons?

Sorry, I was confused, I put it in the wrong context.

Reply:
So what is the conclusion on this? Does it require a change in the code?

Reply:
@jfischer-no @desowin ping

Reply:
@palhaland The same applies to the line you linked to. I don't really know the reason behind it and would like the `USB_TRANS_NO_ZLP` be removed for `USB_TRANS_READ` for clarity reasons.

Reply:
If @jfischer-no is ok with the change in bluetooth.c, then I don't mind doing that change as a part of this pull request

Reply:
I opened #53532 since I was there anyway as I tested another PR.